core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -161,31 +205,32 @@ public void close() {
    */
   private void flushProc() {
     /**
-     * Runs the loop until thread is interrupted.
+     * Runs the loop until ::stop() is called.
      */
-    boolean exit = false;
-    while (!exit) {
+    while (!mStopFlushing) {
 
       /**
-       * Wait to be woken up by flush requests as long as the queue is empty. Bail and drain the
-       * queue, if interrupted.
+       * Stand still unless;
+       * - queue has items
+       * - permit is given by:
+       *   - clients
+       *   -::stop()
        */
-      synchronized (mFlushThread) {
-        while (mQueue.isEmpty()) {
-          try {
-            mFlushThread.wait();
-          } catch (InterruptedException ie) {
-            // Time to exit.
-            exit = true;
+      while (mQueue.isEmpty() && !mStopFlushing) {
+        try {
+          // Wait for permit up to batch timeout.
+          // PS: We don't wait for permit indefinitely in order to process
+          // queued entries proactively.
+          if (mFlushSemaphore.tryAcquire(mFlushBatchTimeNs, TimeUnit.NANOSECONDS)) {
```

Comment:
Using semaphore because I didn't want the signal to be lost, when signaling is done before waiting for signal. Lost signals introduce latency issue and I couldn't find a monitor condition that I can rely on.

Reply:
The client threads will release many times, but this thread will only acquire 1 at a time. Is that an issue? It would basically wake up the first time, process all the tickets, then keep acquiring and doing nothing.

Reply:
As the flush thread process tickets, ticket holders will reacquire the permit that they have given. If all tickets have been processed, then the thread will find no permit. If some tickets have not been processed, then remaining permits will allow flush thread to go into processing again.

There is still a windows where flush thread might acquire the permit before a processed ticket holder do. In that case, it may recycle unnecessarily. However, since we process tickets serially, that window is very narrow and it shouldn't allow thread to recycle unnecessarily more than couple times even under heavy load.